File: src/fmt/Backtracking/CombinationSum.java
package fmt.Backtracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Given a set of candidate numbers (candidates) (without duplicates) and a target number (target), find all unique
 * combinations in candidates where the candidate numbers sums to target.
 * <p>
 * The same repeated number may be chosen from candidates unlimited number of times.
 * <p>
 * Note:
 * <p>
 * All numbers (including target) will be positive integers. The solution set must not contain duplicate combinations.
 * Example 1:
 * <p>
 * Input: candidates = [2,3,6,7], target = 7, A solution set is: [ [7], [2,2,3] ] Example 2:
 * <p>
 * Input: candidates = [2,3,5], target = 8, A solution set is: [ [2,2,2,2], [2,3,3], [3,5] ]
 */
// DFS, Backtracking
public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);

        backtrack(res, new ArrayList<>(), nums, target, 0);

        return res;
    }

    private void backtrack(List<List<Integer>> list, List<Integer> tempList, int[] nums, int remain, int start) {
        if (remain < 0) return;
        else if (remain == 0) list.add(new ArrayList<>(tempList));
        else {
            for (int i = start; i < nums.length; i++) {
                tempList.add(nums[i]);
                backtrack(list, tempList, nums, remain - nums[i], i); // not i + 1 because we can reuse same elements
                tempList.remove(tempList.size() - 1);
            }
        }
    }

    public static void main(String[] args) {

    }
}

File: src/dsa/preface/js/CombinationSum.js
const getNums = (nums, target) => {
  nums.sort((a, b) => a - b); // nlogn
  const res = [];
  backtrack(res, nums, 0, target, 0, []);

  return res;
};


const backtrack = (res, nums, start, target, current, list) => {
  if (target < current) return;
  if (target === current) {
    res.push([...list]); // clone of list
    return;
  } else {
    for (let i = start; i < nums.length; i++) {

      list.push(nums[i]);

      backtrack(res, nums, i + 1, target, current + nums[i], list);

      list.pop();
    }
  }
};

const nums = [1, 2, 3];
const target = 3;
console.log(getNums(nums, target));